ast/parser: Detect function rule head + contains keyword.

philipaconrad opened the pull request:
Previously, we did not detect misuses of the `contains` keyword, which is only valid as a syntactic sugar for partial set definition rules.

This PR adds some logic to detect when a function rule head has been paired with the `contains` keyword, which should generally be an error condition.

~~Note: We can't catch the following case, because the number of args is zero, and we can't distinguish `nil` from `[]` in Go.~~
EDIT: I was a smidge tired when I made the initial PR, and mixed up the nil rules in Golang around slices. It's `len()` that doesn't distinguish between `nil` and `[]`. Silly mistakes happen, even to the best of us. :sweat_smile: (Thanks for the correction, @srenatus!)

```rego
package test

# Both of these cases result in an error condition:
p() contains x {
    x := 1
}

q(a) contains x {
    x := a
}
```

Fixes: #5525

philipaconrad commented on ast/compile_test.go:
```diff
@@ -1734,6 +1734,24 @@ p {
 	}
 }
 
+func TestCompilerCheckSafetyFunctionAndContainsKeyword(t *testing.T) {
```
Originally, the check lived under `check.go`, and so the compiler was doing the error checking. I left the extra safety check in, since it's not hurting anything, and it's nice to have a downstream safety check.

philipaconrad commented on ast/parser.go:
```diff
@@ -831,6 +831,10 @@ func (p *Parser) parseHead(defaultRule bool) (*Head, bool) {
 
 	switch p.s.tok {
 	case tokens.Contains: // NOTE: no Value for `contains` heads, we return here
+		// Catch error case of using 'contains' with a function definition rule head.
```
This is located before the scan to have the error location land _on_ the keyword. Using `p.save()` + `p.restore()` later would have a similar effect.

philipaconrad commented on ast/check.go:
```diff
@@ -213,7 +213,6 @@ func (tc *typeChecker) checkRule(env *TypeEnv, as *AnnotationSet, rule *Rule) {
 	var tpe types.Type
 
 	if len(rule.Head.Args) > 0 {
```
This is a minor diff that got left in after a local rebase (when the check moved from `check.go` -> `parser.go`).

srenatus commented:
> we can't distinguish nil from [] in Go

We can't? I believe only `len()` on a slice can't, but we could do it differently. [See this playground.](https://play.golang.com/p/6QefE0-dJyC)

srenatus reviewed (approved):
Thanks! Nitpicks only. 👍

philipaconrad commented:
> I just got surprised by the statement in isolation

@srenatus You were right to be surprised! Because I was *wrong*, and the PR has wound up better for the correction. :sweat_smile: For some reason, I had gotten confused about the nil rules in Go, and that resulted in a less-accurate check.

I've fixed the goofy logic, and added an extra test case to check for the "function with 0 arguments" case.

philipaconrad commented on ast/compile_test.go:
```diff
@@ -1734,6 +1734,24 @@ p {
+
+	errs := err.(Errors)
+	if !strings.Contains(errs[0].Message, "the contains keyword can only be used with partial set definitions (e.g., p contains <VALUE> { ... })") || errs[0].Location.Row != 5 {
+		t.Fatal("expected error on row 5 but got:", err)
```
This is fixed in the latest changeset by splitting the row and message content checks apart. :+1:

philipaconrad commented on ast/parser.go:
```diff
@@ -831,6 +831,10 @@ func (p *Parser) parseHead(defaultRule bool) (*Head, bool) {
 	case tokens.Contains: // NOTE: no Value for `contains` heads, we return here
+		// Catch error case of using 'contains' with a function definition rule head.
+		if len(head.Args) > 0 {
+			p.illegal("the contains keyword can only be used with partial set definitions (e.g., %s contains <VALUE> { ... })", name)
```
Thanks for the correction! I noticed the unfamiliar "multi-value rule" language in the error messages, but wasn't exactly sure how to interpret what I was seeing.

I've rolled the language shift into my latest changes.

srenatus reviewed (approved):
👏 🎉 Thanks!
